File: Kotlin/ByRound/2139/2139_B_Cake_Collection.kt
// Problem: CF 2139 B - Cake Collection
// https://codeforces.com/contest/2139/problem/B

/*
B. Cake Collection
Algorithms/Techniques: Greedy algorithm with sorting

Time Complexity: O(n log n + n) where n is the number of ovens, due to sorting and a single pass through the array.
Space Complexity: O(1) excluding input storage, as we only use a constant amount of extra space.

Problem:
Maple wants to bake cakes using magical ovens. Each oven i bakes a_i cakes per second.
At the end of each second, she can collect all cakes from any oven.
We need to maximize the total number of cakes collected in m seconds.
Approach:
- Sort ovens by baking rate in descending order.
- For each of the first min(n,m) seconds, collect from the highest-rated oven available.
- The contribution of an oven is a_i * (m - i), since it's used for (m - i) seconds.
*/

import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.io.StreamTokenizer
import java.util.*
import kotlin.math.min

private val sc = Scanner( System.`in` )
private val br = BufferedReader( InputStreamReader( System.`in` ) )
private val st = StreamTokenizer( br )
private val pw = PrintWriter( System.out )

fun main() {
    solve()
    pw.close()
}

private fun solve() {
    val tn = st.nextInt()
    for ( test in 1 .. tn ) {
        solveTest()
    }
}

private fun solveTest() {
    val n = st.nextInt() // Number of ovens
    val m = st.nextInt() // Number of seconds
    val a = IntArray( n ) { st.nextInt() } // Array of baking rates
    a.sortDescending() // Sort in descending order to prioritize higher rate ovens
    var r = 0L // Result accumulator
    for ( i in 0 until min( n, m ) ) {
        // Multiply the oven's rate by how many seconds it will be used (m - i)
        r += a[i].toLong() * ( m - i )
    }
    pw.println( r )
}

private fun StreamTokenizer.nextInt(): Int {
    nextToken()
    return nval.toInt()
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2139/2139_B_Cake_Collection.py
# Problem: CF 2139 B - Cake Collection
# https://codeforces.com/contest/2139/problem/B

"""
B. Cake Collection
Algorithm: Greedy approach with sorting
- Sort ovens by baking rate in descending order
- For each of the first m seconds, select the oven with highest remaining cake count
- The contribution of an oven at second i is a[i] * (m - i), where i is the index of selection
- Time Complexity: O(n log n) per test case due to sorting
- Space Complexity: O(1) excluding input storage
Techniques:
- Greedy selection: Always pick the oven that will yield maximum cakes in remaining time
- Sorting to prioritize high-yield ovens
"""

for _ in range(int(input())):  # Read number of test cases
    n, m = map(int, input().split())  # Read n (ovens) and m (seconds)
    arr = sorted(map(int, input().split()), reverse=True)  # Sort baking rates in descending order
    ans = 0  # Initialize total cakes collected
    for i in range(min(n, m)):  # Iterate through the first min(n, m) seconds
        ans += arr[i] * (m - i)  # Add contribution of oven i at second i
    print(ans)  # Print maximum cakes collected


# https://github.com/VaHiX/codeForces/